File: backend/app/schemas/intelligence.py
"""统一智能系统 API Schemas — 22 个端点的请求/响应模型

覆盖五大功能域：
1. 会话管理（4）：创建/列表/详情/归档
2. 统一对话（3）：对话/流式/强制模式
3. 上下文与追溯（5）：上下文/追溯/步骤树/成本/决策链
4. 证据收集与分析（4）：项目证据/实体上下文/解读/数据集解读
5. 多模态与规则引擎（6）：标准化/视觉解析/规则列表/详情/执行/验证
"""
from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class BaseSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True, arbitrary_types_allowed=True)


# ========== 1. 会话管理 ==========

class SessionCreate(BaseModel):
    """创建统一会话"""
    title: str = Field("新会话", min_length=1, max_length=200, description="会话标题")
    project_id: Optional[UUID] = Field(None, description="关联项目 ID")
    primary_mode: str = Field("auto", description="主模式: chat/reasoning/agent/auto")


class SessionResponse(BaseSchema):
    """会话响应"""
    id: UUID
    user_id: UUID
    project_id: Optional[UUID] = None
    title: str
    status: str
    primary_mode: str
    context: Optional[Dict[str, Any]] = None
    last_message_at: Optional[datetime] = None
    message_count: int = 0
    created_at: datetime
    updated_at: datetime


class SessionListResponse(BaseSchema):
    """会话列表响应"""
    items: List[SessionResponse]
    total: int


class SessionArchive(BaseModel):
    """归档会话请求"""
    status: str = Field("archived", description="目标状态: archived/deleted")


# ========== 2. 统一对话 ==========

class ChatRequest(BaseModel):
    """统一对话请求"""
    message: str = Field(..., min_length=1, max_length=10000, description="用户消息")
    project_id: Optional[UUID] = Field(None, description="项目 ID（覆盖会话）")
    force_mode: Optional[str] = Field(None, description="强制模式: chat/reasoning/agent/hybrid")
    capability_hint: Optional[str] = Field(None, description="能力提示: qa/reasoning/agent/auto（Agent网关使用）")
    tier: Optional[str] = Field(None, description="档位: turbo/standard/deep,None 或 auto 时智能推荐")


class ChatResponse(BaseSchema):
    """统一对话响应"""
    answer: str = Field("", description="助手回复")
    mode: str = Field("chat", description="实际路由模式")
    intent: Optional[Dict[str, Any]] = None
    session_id: str
    cost_usd: float = 0.0
    duration_sec: float = 0.0
    tier: str = Field("standard", description="实际使用的档位")
    tier_reason: Optional[str] = Field(None, description="档位选择原因")


class TierSuggestRequest(BaseModel):
    """档位推荐请求"""
    message: str = Field(..., min_length=1, max_length=5000, description="用户消息")


class TierSuggestResponse(BaseSchema):
    """档位推荐响应"""
    tier: str = Field(..., description="推荐档位 turbo/standard/deep")
    reason: str = Field("", description="推荐原因")
    confidence: float = Field(0.0, description="意图置信度 0-1")
    tier_config: Dict[str, Any] = Field(default_factory=dict, description="档位配置")


class ForceModeRequest(BaseModel):
    """强制切换模式请求"""
    mode: str = Field(..., description="目标模式: chat/reasoning/agent/hybrid")


# ========== 3. 上下文与追溯 ==========

class ContextMemoryItem(BaseSchema):
    """上下文记忆条目"""
    id: UUID
    type: str
    content: Any
    importance: float = 0.5
    created_at: Optional[datetime] = None


class ContextResponse(BaseSchema):
    """上下文响应"""
    session_id: str
    memories: List[ContextMemoryItem]
    context_prompt: str = ""


class TraceStep(BaseSchema):
    """推理步骤"""
    id: UUID
    step_type: str
    agent_name: Optional[str] = None
    phase: Optional[str] = None
    round_num: Optional[int] = None
    decision_basis: Optional[str] = None
    cost_usd: Optional[float] = None
    duration_sec: Optional[float] = None
    status: str = "completed"
    created_at: Optional[datetime] = None
    evidence: Optional[Dict[str, Any]] = None


class TraceResponse(BaseSchema):
    """推理追溯响应"""
    session_id: str
    total_steps: int
    traces: List[TraceStep]


class TraceTreeResponse(BaseSchema):
    """步骤树响应"""
    roots: List[Dict[str, Any]]
    total_steps: int
    total_cost: float = 0.0


class CostBreakdownResponse(BaseSchema):
    """成本分解响应"""
    total_cost: float = 0.0
    total_tokens: int = 0
    by_agent: Dict[str, float] = Field(default_factory=dict)
    by_phase: Dict[str, float] = Field(default_factory=dict)
    by_step_type: Dict[str, float] = Field(default_factory=dict)


class DecisionChainResponse(BaseSchema):
    """决策链响应"""
    decisions: List[Dict[str, Any]]


# ========== 4. 证据收集与分析 ==========

class EvidenceCollectRequest(BaseModel):
    """证据收集请求"""
    project_id: Optional[UUID] = Field(None, description="项目 ID")
    trigger_event: Optional[str] = Field(None, description="触发事件类型")
    entity_id: Optional[UUID] = Field(None, description="触发实体 ID")
    extra_evidence: Optional[str] = Field(None, description="额外证据文本")


class EvidenceSourceItem(BaseSchema):
    """证据来源条目"""
    source_type: str
    count: int
    detail: str = ""
    snippets_kept: List[str] = Field(default_factory=list, description="保留的top-3命中片段原文（限长100字）")


class EvidenceResponse(BaseSchema):
    """证据收集响应"""
    text: str = ""
    sources: List[EvidenceSourceItem] = Field(default_factory=list)
    total_items: int = 0
    project_id: Optional[str] = None
    entity_id: Optional[str] = None
    trigger_event: Optional[str] = None


class AnalysisInterpretRequest(BaseModel):
    """解读请求"""
    message: str = Field(..., min_length=1, description="分析目标/问题")
    analysis_data: Optional[Dict[str, Any]] = Field(None, description="已有分析数据")
    project_id: Optional[UUID] = Field(None, description="项目 ID")
    intent: Optional[str] = Field(None, description="指定意图")


class AnalysisInterpretResponse(BaseSchema):
    """解读响应"""
    intent: str
    conclusion: str
    hypothesis: str
    recommendations: List[str] = Field(default_factory=list)
    key_findings: List[str] = Field(default_factory=list)
    model: str = ""
    cost_usd: float = 0.0
    duration_sec: float = 0.0
    evidence_summary: Optional[Dict[str, Any]] = None


class DatasetInterpretRequest(BaseModel):
    """数据集解读请求"""
    message: Optional[str] = Field(None, description="解读方向（可选）")
    project_id: Optional[UUID] = Field(None, description="项目 ID（补充上下文）")


# ========== 5. 多模态与规则引擎 ==========

class MultimodalNormalizeRequest(BaseModel):
    """多模态标准化请求"""
    text: Optional[str] = Field(None, description="文本输入")
    image_paths: Optional[List[str]] = Field(None, description="本地图像文件路径列表")
    image_urls: Optional[List[str]] = Field(None, description="图像 URL 列表")
    image_base64: Optional[List[str]] = Field(None, description="图像 base64 编码列表")
    file_paths: Optional[List[str]] = Field(None, description="文件路径列表")
    structured_data: Optional[Dict[str, Any]] = Field(None, description="结构化数据")


class MultimodalNormalizeResponse(BaseSchema):
    """多模态标准化响应"""
    items: List[Dict[str, Any]] = Field(default_factory=list)
    primary_text: str = ""
    has_image: bool = False
    modalities: List[str] = Field(default_factory=list)
    textualized: str = ""


class VisionAnalyzeRequest(BaseModel):
    """视觉解析请求"""
    image_data_uri: str = Field(..., description="图像数据 URI 或 URL")
    prompt: str = Field(..., description="分析提示词")
    analysis_type: Optional[str] = Field(
        None, description="分析类型: pathology/protein_structure/molecule_structure/chart/general"
    )
    focus: Optional[str] = Field(None, description="关注焦点")


class VisionAnalyzeResponse(BaseSchema):
    """视觉解析响应"""
    description: str
    model: str = ""
    usage: Dict[str, Any] = Field(default_factory=dict)
    cost_usd: float = 0.0
    duration_sec: float = 0.0


class RuleResponse(BaseSchema):
    """规则详情响应"""
    id: str
    name: str
    when: Dict[str, Any]
    then: List[Dict[str, Any]]
    priority: int = 0
    enabled: bool = True
    description: Optional[str] = None
    tags: Optional[List[str]] = None


class RuleSetResponse(BaseSchema):
    """规则集响应"""
    name: str
    version: str = "1.0"
    description: Optional[str] = None
    rules: List[RuleResponse] = Field(default_factory=list)


class RuleListResponse(BaseSchema):
    """规则列表响应"""
    presets: List[str] = Field(default_factory=list)
    rulesets: List[RuleSetResponse] = Field(default_factory=list)
    total_rules: int = 0


class RuleExecuteRequest(BaseModel):
    """规则执行请求"""
    preset: Optional[str] = Field(None, description="内置 preset 名称")
    yaml_content: Optional[str] = Field(None, description="自定义 YAML 内容（与 preset 二选一）")
    context: Dict[str, Any] = Field(..., description="求值上下文")
    tags: Optional[List[str]] = Field(None, description="仅执行含指定标签的规则")


class RuleExecutionResultItem(BaseSchema):
    """单条规则执行结果"""
    rule_id: str
    rule_name: str
    matched: bool
    actions_executed: int = 0
    outputs: List[Dict[str, Any]] = Field(default_factory=list)
    error: Optional[str] = None


class RuleExecuteResponse(BaseSchema):
    """规则执行响应"""
    ruleset_name: str
    total_rules: int
    matched_rules: int
    executed_actions: int
    results: List[RuleExecutionResultItem]
    context_changes: Dict[str, Any] = Field(default_factory=dict)
    duration_sec: float = 0.0


class RuleValidateRequest(BaseModel):
    """规则验证请求"""
    yaml_content: str = Field(..., description="待验证的 YAML 内容")


class RuleValidateResponse(BaseSchema):
    """规则验证响应"""
    valid: bool
    errors: List[str] = Field(default_factory=list)
    rules_count: int = 0
    ruleset_name: Optional[str] = None


__all__ = [
    # 会话管理
    "SessionCreate", "SessionResponse", "SessionListResponse", "SessionArchive",
    # 统一对话
    "ChatRequest", "ChatResponse", "TierSuggestRequest", "TierSuggestResponse", "ForceModeRequest",
    # 上下文与追溯
    "ContextMemoryItem", "ContextResponse", "TraceStep", "TraceResponse",
    "TraceTreeResponse", "CostBreakdownResponse", "DecisionChainResponse",
    # 证据收集与分析
    "EvidenceCollectRequest", "EvidenceSourceItem", "EvidenceResponse",
    "AnalysisInterpretRequest", "AnalysisInterpretResponse",
    "DatasetInterpretRequest",
    # 多模态与规则引擎
    "MultimodalNormalizeRequest", "MultimodalNormalizeResponse",
    "VisionAnalyzeRequest", "VisionAnalyzeResponse",
    "RuleResponse", "RuleSetResponse", "RuleListResponse",
    "RuleExecuteRequest", "RuleExecuteResponse",
    "RuleValidateRequest", "RuleValidateResponse",
]

File: frontend/lib/api/intelligence.ts
/**
 * 统一智能系统（Intelligence）API 客户端
 *
 * 对应后端 app/api/v1/endpoints/intelligence.py 的 22 个端点。
 *
 * 五大功能域：
 *  1. 会话管理（createSession / listSessions / getSession / archiveSession）
 *  2. 统一对话（chat / streamChat / forceMode）— streamChat 使用 fetch + ReadableStream 解析 SSE
 *  3. 上下文与追溯（getContext / getTrace / getTraceTree / getCostBreakdown / getDecisionChain）
 *  4. 证据收集与分析（collectEvidence / collectEntityContext / interpretAnalysis / interpretDataset）
 *  5. 多模态与规则引擎（normalizeMultimodal / analyzeVision / listRules / getRulePreset / executeRules / validateRules）
 */
import { api } from './client';

// ============================================================================
// 类型定义（与后端 app/schemas/intelligence.py 对齐）
// ============================================================================

export type PrimaryMode = 'chat' | 'reasoning' | 'agent' | 'hybrid' | 'auto';

export interface SessionResponse {
  id: string;
  user_id: string;
  project_id?: string | null;
  title: string;
  status: string;
  primary_mode: PrimaryMode;
  context?: Record<string, unknown> | null;
  last_message_at?: string | null;
  message_count: number;
  created_at?: string;
  updated_at?: string;
}

export interface SessionListResponse {
  items: SessionResponse[];
  total: number;
}

export interface SessionCreatePayload {
  title?: string;
  project_id?: string;
  primary_mode?: PrimaryMode;
}

export interface SessionArchivePayload {
  status: 'archived' | 'deleted';
}

export interface ChatRequest {
  message: string;
  project_id?: string;
  force_mode?: PrimaryMode;
}

export interface ChatResponse {
  answer: string;
  mode: string;
  [key: string]: unknown;
}

export interface ContextMemoryItem {
  id: string;
  type: string;
  content: unknown;
  importance: number;
  created_at?: string | null;
}

export interface ContextResponse {
  session_id: string;
  memories: ContextMemoryItem[];
  context_prompt: string;
}

export interface TraceStep {
  id: string;
  step_type: string;
  agent_name?: string | null;
  phase?: string | null;
  round_num?: number | null;
  decision_basis?: string | null;
  cost_usd?: number | null;
  duration_sec?: number | null;
  status: string;
  created_at?: string | null;
}

export interface TraceResponse {
  session_id: string;
  total_steps: number;
  traces: TraceStep[];
}

export interface TraceTreeResponse {
  roots: Array<Record<string, unknown>>;
  total_steps: number;
  total_cost: number;
}

export interface CostBreakdownResponse {
  total_cost: number;
  total_tokens: number;
  by_agent: Record<string, number>;
  by_phase: Record<string, number>;
  by_step_type: Record<string, number>;
}

export interface DecisionChainResponse {
  decisions: Array<Record<string, unknown>>;
}

export interface StreamCallbacks {
  onChunk?: (chunk: string) => void;
  onDone?: (fullText: string) => void;
  onError?: (err: Error) => void;
  signal?: AbortSignal;
}

export interface EvidenceCollectPayload {
  project_id?: string;
  entity_id?: string;
  trigger_event?: string;
  extra_evidence?: Record<string, unknown>;
}

export interface EvidenceResponse {
  text: string;
  sources: Array<Record<string, unknown>>;
  total_items: number;
  project_id?: string | null;
  entity_id?: string | null;
  trigger_event?: string | null;
}

export interface AnalysisInterpretPayload {
  message: string;
  analysis_data?: Record<string, unknown>;
  project_id?: string;
  intent?: string;
}

export interface DatasetInterpretPayload {
  message: string;
  project_id?: string;
}

export interface AnalysisInterpretResponse {
  interpretation?: string;
  suggestions?: string[];
  [key: string]: unknown;
}

export interface MultimodalNormalizePayload {
  text?: string;
  image_paths?: string[];
  image_urls?: string[];
  image_base64?: string[];
  file_paths?: string[];
  structured_data?: Record<string, unknown>;
}

export interface MultimodalNormalizeResponse {
  items: Array<Record<string, unknown>>;
  primary_text: string;
  has_image: boolean;
  modalities: string[];
  textualized: string;
}

export interface VisionAnalyzePayload {
  image_data_uri: string;
  analysis_type?: 'pathology' | 'protein_structure' | 'molecule_structure' | 'chart';
  prompt?: string;
  focus?: string;
}

export interface VisionAnalyzeResponse {
  [key: string]: unknown;
}

export interface RuleResponseItem {
  id: string;
  name: string;
  when: Record<string, unknown>;
  then: Array<Record<string, unknown>>;
  priority: number;
  enabled: boolean;
  description?: string;
  tags?: string[];
}

export interface RuleSetResponse {
  name: string;
  version: string;
  description?: string;
  rules: RuleResponseItem[];
}

export interface RuleListResponse {
  presets: string[];
  rulesets: RuleSetResponse[];
  total_rules: number;
}

export interface RuleExecutePayload {
  preset?: string;
  yaml_content?: string;
  context: Record<string, unknown>;
  tags?: string[];
}

export interface RuleExecuteResponse {
  ruleset_name: string;
  total_rules: number;
  matched_rules: number;
  executed_actions: number;
  results: Array<Record<string, unknown>>;
  context_changes: Record<string, unknown>;
  duration_sec: number;
}

export interface RuleValidatePayload {
  yaml_content: string;
}

export interface RuleValidateResponse {
  valid: boolean;
  errors: string[];
  rules_count: number;
  ruleset_name?: string | null;
}

// ============================================================================
// 内部工具
// ============================================================================

/** 解析 axios 响应已解包的 data 字段（响应拦截器自动拆信封） */
const unwrap = <T,>(resp: { data: unknown }): T => resp.data as T;

/** 构建 fetch URL（复用 client.ts 的 baseURL 规则） */
const API_BASE =
  (typeof process !== 'undefined' && process.env?.NEXT_PUBLIC_API_BASE) ||
  'http://localhost:8000/api/v1';

const buildUrl = (path: string): string => {
  const base = API_BASE.endsWith('/') ? API_BASE.slice(0, -1) : API_BASE;
  const cleanPath = path.startsWith('/') ? path : `/${path}`;
  return `${base}${cleanPath}`;
};

/** 从 localStorage 读取 JWT（与 client.ts 保持一致） */
const readAuthToken = (): string | null => {
  if (typeof window === 'undefined') return null;
  return window.localStorage.getItem('ai_drug_token');
};

// ============================================================================
// 1. 会话管理
// ============================================================================

export const createSession = (payload: SessionCreatePayload = {}): Promise<SessionResponse> =>
  api.post('/intelligence/sessions', payload).then(unwrap<SessionResponse>);

export const listSessions = (params: {
  project_id?: string;
  limit?: number;
} = {}): Promise<SessionListResponse> =>
  api.get('/intelligence/sessions', { params }).then(unwrap<SessionListResponse>);

export const getSession = (sessionId: string): Promise<SessionResponse> =>
  api.get(`/intelligence/sessions/${sessionId}`).then(unwrap<SessionResponse>);

export const archiveSession = (
  sessionId: string,
  payload: SessionArchivePayload = { status: 'archived' }
): Promise<SessionResponse> =>
  api.patch(`/intelligence/sessions/${sessionId}`, payload).then(unwrap<SessionResponse>);

// ============================================================================
// 2. 统一对话（非流式 + 流式）
// ============================================================================

/** 非流式 POST /intelligence/sessions/{sessionId}/chat */
export const sendChat = (
  sessionId: string,
  message: string,
  options: { projectId?: string; forceMode?: PrimaryMode } = {}
): Promise<ChatResponse> => {
  const body: ChatRequest = { message };
  if (options.projectId) body.project_id = options.projectId;
  if (options.forceMode) body.force_mode = options.forceMode;
  return api
    .post(`/intelligence/sessions/${sessionId}/chat`, body)
    .then(unwrap<ChatResponse>);
};

/** 强制切换主模式 */
export const forceMode = (sessionId: string, mode: PrimaryMode): Promise<{ primary_mode: string }> =>
  api
    .post(`/intelligence/sessions/${sessionId}/force-mode`, { mode })
    .then(unwrap<{ primary_mode: string }>);

/**
 * 流式对话（SSE）
 *
 * 由于 EventSource 仅支持 GET，这里使用 fetch + ReadableStream 解析 `data: xxx\n\n` 帧。
 * 后端会输出 `data: [DONE]\n\n` 作为结束帧，`data: [ERROR] xxx\n\n` 作为异常帧。
 */
export const streamChat = async (
  sessionId: string,
  message: string,
  options: {
    projectId?: string;
    forceMode?: PrimaryMode;
    onChunk?: (chunk: string) => void;
    onDone?: (fullText: string) => void;
    onError?: (err: Error) => void;
    signal?: AbortSignal;
  } = {}
): Promise<string> => {
  const url = buildUrl(`/intelligence/sessions/${sessionId}/stream`);

  const body: ChatRequest = { message };
  if (options.projectId) body.project_id = options.projectId;
  if (options.forceMode) body.force_mode = options.forceMode;

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    Accept: 'text/event-stream',
  };
  const token = readAuthToken();
  if (token) headers['Authorization'] = `Bearer ${token}`;

  let response: Response;
  try {
    response = await fetch(url, {
      method: 'POST',
      headers,
      body: JSON.stringify(body),
      signal: options.signal,
      cache: 'no-store',
    });
  } catch (err) {
    const error = err instanceof Error ? err : new Error(String(err));
    options.onError?.(error);
    throw error;
  }

  if (!response.ok || !response.body) {
    const text = await response.text().catch(() => '');
    const error = new Error(
      `SSE 请求失败：HTTP ${response.status} ${response.statusText} ${text}`.trim()
    );
    options.onError?.(error);
    throw error;
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder('utf-8');
  let buffer = '';
  let fullText = '';

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });

      let sepIdx: number;
      while ((sepIdx = buffer.indexOf('\n\n')) !== -1) {
        const rawEvent = buffer.slice(0, sepIdx);
        buffer = buffer.slice(sepIdx + 2);

        const lines = rawEvent.split('\n');
        for (const line of lines) {
          const trimmed = line.trim();
          if (!trimmed.startsWith('data:')) continue;
          const data = trimmed.slice(5).trim();

          if (data === '[DONE]') {
            options.onDone?.(fullText);
            return fullText;
          }
          if (data.startsWith('[ERROR]')) {
            const errMsg = data.slice('[ERROR]'.length).trim() || '未知流式错误';
            const error = new Error(errMsg);
            options.onError?.(error);
            throw error;
          }

          fullText += data;
          options.onChunk?.(data);
        }
      }
    }

    if (buffer.trim().length > 0) {
      const lines = buffer.split('\n');
      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed.startsWith('data:')) continue;
        const data = trimmed.slice(5).trim();
        if (data === '[DONE]') {
          options.onDone?.(fullText);
          return fullText;
        }
        if (data.startsWith('[ERROR]')) {
          const errMsg = data.slice('[ERROR]'.length).trim() || '未知流式错误';
          const error = new Error(errMsg);
          options.onError?.(error);
          throw error;
        }
        fullText += data;
        options.onChunk?.(data);
      }
    }

    options.onDone?.(fullText);
    return fullText;
  } catch (err) {
    if (err instanceof Error && err.name === 'AbortError') {
      throw err;
    }
    const error = err instanceof Error ? err : new Error(String(err));
    options.onError?.(error);
    throw error;
  } finally {
    try {
      reader.releaseLock?.();
    } catch {
      /* noop */
    }
  }
};

// ============================================================================
// 3. 上下文与追溯
// ============================================================================

export const getContext = (sessionId: string, limit = 50): Promise<ContextResponse> =>
  api
    .get(`/intelligence/sessions/${sessionId}/context`, { params: { limit } })
    .then(unwrap<ContextResponse>);

export const getTrace = (sessionId: string, limit = 200): Promise<TraceResponse> =>
  api
    .get(`/intelligence/sessions/${sessionId}/trace`, { params: { limit } })
    .then(unwrap<TraceResponse>);

export const getTraceTree = (runId: string): Promise<TraceTreeResponse> =>
  api.get(`/intelligence/runs/${runId}/trace-tree`).then(unwrap<TraceTreeResponse>);

export const getCostBreakdown = (runId: string): Promise<CostBreakdownResponse> =>
  api.get(`/intelligence/runs/${runId}/cost`).then(unwrap<CostBreakdownResponse>);

export const getDecisionChain = (runId: string): Promise<DecisionChainResponse> =>
  api.get(`/intelligence/runs/${runId}/decisions`).then(unwrap<DecisionChainResponse>);

// ============================================================================
// 4. 证据收集与分析
// ============================================================================

export const collectEvidence = (payload: EvidenceCollectPayload): Promise<EvidenceResponse> =>
  api.post('/intelligence/evidence/collect', payload).then(unwrap<EvidenceResponse>);

export const collectEntityContext = (payload: EvidenceCollectPayload): Promise<EvidenceResponse> =>
  api.post('/intelligence/evidence/collect-entity', payload).then(unwrap<EvidenceResponse>);

export const interpretAnalysis = (payload: AnalysisInterpretPayload): Promise<AnalysisInterpretResponse> =>
  api
    .post('/intelligence/analysis/interpret', payload)
    .then(unwrap<AnalysisInterpretResponse>);

export const interpretDataset = (
  datasetId: string,
  payload: DatasetInterpretPayload
): Promise<AnalysisInterpretResponse> =>
  api
    .post(`/intelligence/analysis/datasets/${datasetId}/interpret`, payload)
    .then(unwrap<AnalysisInterpretResponse>);

// ============================================================================
// 5. 多模态与规则引擎
// ============================================================================

export const normalizeMultimodal = (
  payload: MultimodalNormalizePayload
): Promise<MultimodalNormalizeResponse> =>
  api
    .post('/intelligence/multimodal/normalize', payload)
    .then(unwrap<MultimodalNormalizeResponse>);

export const analyzeVision = (payload: VisionAnalyzePayload): Promise<VisionAnalyzeResponse> =>
  api.post('/intelligence/vision/analyze', payload).then(unwrap<VisionAnalyzeResponse>);

export const listRules = (): Promise<RuleListResponse> =>
  api.get('/intelligence/rules').then(unwrap<RuleListResponse>);

export const getRulePreset = (preset: string): Promise<RuleSetResponse> =>
  api.get(`/intelligence/rules/${preset}`).then(unwrap<RuleSetResponse>);

export const executeRules = (payload: RuleExecutePayload): Promise<RuleExecuteResponse> =>
  api.post('/intelligence/rules/execute', payload).then(unwrap<RuleExecuteResponse>);

export const validateRules = (payload: RuleValidatePayload): Promise<RuleValidateResponse> =>
  api.post('/intelligence/rules/validate', payload).then(unwrap<RuleValidateResponse>);
